File: 백준/Silver/3273. 두 수의 합/두 수의 합.js
const fs = require('fs');
// 백준 제출용
const input = fs.readFileSync(0).toString().trim().split('\n');

// 로컬 테스트용
// const filePath =
//   process.platform === 'linux' ? 'dev/stdin' : __dirname + '/input.txt';
// let input = fs.readFileSync(filePath).toString().trim().split('\n');

let n = Number(input[0]);
let arr = input[1].split(' ').map(Number);
let x = Number(input[2]);

//console.log(arr);
arr.sort((a, b) => a - b);

let start = 0;
let end = n - 1;
let count = 0; // 조건을 만족하는 쌍의 개수
let curSum = 0;

/*
 * two-pointer 진행
 */
while (start < end) {
  // 현재 x값 선언
  curSum = arr[start] + arr[end];

  // 타겟값보다 작으면
  if (curSum < x) {
    start++; // 오른쪽으로 이동
  } // 타겟값보다 크면
  else if (curSum > x) {
    end--; // 왼쪽으로 이동
  }
  // 타켓값과 일치 => 조건을 만족하는 경우
  else {
    count++;
    start++;
    end--;
  }
}

console.log(count);

File: 백준/Silver/3273. 두 수의 합/두 수의 합.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int n = Integer.parseInt(st.nextToken());
		int[] data = new int[n]; // 수열 배열

		// 수열 내 원소 입력 처리
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++) {
			data[i] = Integer.parseInt(st.nextToken());
		}

		int x = Integer.parseInt(br.readLine()); // target: x

		Arrays.sort(data);

		int start = 0; // 포인터1
		int end = n - 1; // 포인터2
		int count = 0; // 조건을 만족하는 쌍의 개수
		int curSum = 0; // 현재 포인터의 값 합계

		// two-pointer 진행
		while (start < end) {
			curSum = data[start] + data[end];

			// 타겟값보다 작으면
			if (curSum < x) {
				start++; // 오른쪽으로 이동
			} // 타겟값보다 크면
			else if (curSum > x) {
				end--; // 왼쪽으로 이동
			} else {
				// 타켓값과 일치 => 조건을 만족하는 경우
				count++;
				start++; // 다음 값 찾기
				end--; // 다음 값 찾기
			}
		}

		System.out.println(count); // 결과 출력
	}
}